File: backend/src/main/kotlin/com/tripbook/catalog/service/ReviewService.kt
package com.tripbook.catalog.service

import com.tripbook.catalog.model.Review
import com.tripbook.catalog.model.ReviewSummary
import java.util.UUID

/**
 * Service interface for business logic related to review management.
 * Defines methods for CRUD operations and queries related to company reviews.
 */
interface ReviewService {
    /**
     * Retrieves all reviews for a specific company.
     * 
     * @param companyId The ID of the company
     * @return List of reviews for the company
     */
    suspend fun getReviewsByCompany(companyId: UUID): List<Review>
    
    /**
     * Retrieves a specific review by ID.
     * 
     * @param id The review ID
     * @return The review if found, null otherwise
     */
    suspend fun getReviewById(id: UUID): Review?
    
    /**
     * Creates a new review.
     * 
     * @param review The review to create
     * @return The created review
     */
    suspend fun createReview(review: Review): Review
    
    /**
     * Updates an existing review.
     * 
     * @param id The ID of the review to update
     * @param review The updated review data
     * @return The updated review if found, null otherwise
     */
    suspend fun updateReview(id: UUID, review: Review): Review?
    
    /**
     * Deletes a review by ID.
     * 
     * @param id The ID of the review to delete
     * @return true if deleted, false if not found
     */
    suspend fun deleteReview(id: UUID): Boolean
    
    /**
     * Marks a review as helpful.
     * 
     * @param reviewId The ID of the review
     * @param userId The ID of the user marking the review as helpful
     * @return true if successful, false otherwise
     */
    suspend fun markReviewAsHelpful(reviewId: UUID, userId: UUID): Boolean
    
    /**
     * Retrieves the review summary for a company.
     * 
     * @param companyId The ID of the company
     * @return The review summary with statistics
     */
    suspend fun getReviewSummaryForCompany(companyId: UUID): ReviewSummary
    
    /**
     * Retrieves all reviews submitted by a specific user.
     * 
     * @param userId The ID of the user
     * @return List of reviews by the user
     */
    suspend fun getReviewsByUser(userId: UUID): List<Review>
}

File: lib/services/review_service.dart
// Flutter service for review-related API calls
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/review.dart';
import '../utils/api_config.dart';

/// Service class for review-related API operations
class ReviewService {
  final http.Client _httpClient;
  final String _baseUrl;
  
  ReviewService({
    http.Client? httpClient,
    String? baseUrl,
  }) : _httpClient = httpClient ?? http.Client(),
       _baseUrl = baseUrl ?? ApiConfig.baseUrl;
  
  /// Get reviews for a company
  Future<List<Review>> getReviewsByCompany(String companyId) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/reviews/company/$companyId'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Review.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load reviews: ${response.body}');
    }
  }
  
  /// Get review by ID
  Future<Review> getReviewById(String id) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/reviews/$id'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      return Review.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load review: ${response.body}');
    }
  }
  
  /// Submit a new review
  Future<Review> createReview(Review review) async {
    final response = await _httpClient.post(
      Uri.parse('$_baseUrl/api/reviews'),
      headers: ApiConfig.headers,
      body: json.encode(review.toJson()),
    );
    
    if (response.statusCode == 201) {
      return Review.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to create review: ${response.body}');
    }
  }
  
  /// Update an existing review
  Future<Review> updateReview(String id, Review review) async {
    final response = await _httpClient.put(
      Uri.parse('$_baseUrl/api/reviews/$id'),
      headers: ApiConfig.headers,
      body: json.encode(review.toJson()),
    );
    
    if (response.statusCode == 200) {
      return Review.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to update review: ${response.body}');
    }
  }
  
  /// Delete a review
  Future<bool> deleteReview(String id) async {
    final response = await _httpClient.delete(
      Uri.parse('$_baseUrl/api/reviews/$id'),
      headers: ApiConfig.headers,
    );
    
    return response.statusCode == 204;
  }
  
  /// Mark a review as helpful
  Future<bool> markReviewAsHelpful(String reviewId, String userId) async {
    final response = await _httpClient.post(
      Uri.parse('$_baseUrl/api/reviews/$reviewId/helpful?userId=$userId'),
      headers: ApiConfig.headers,
    );
    
    return response.statusCode == 200;
  }
  
  /// Get review summary for a company
  Future<ReviewSummary> getReviewSummary(String companyId) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/reviews/summary/$companyId'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      return ReviewSummary.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load review summary: ${response.body}');
    }
  }
  
  /// Get reviews by a user
  Future<List<Review>> getReviewsByUser(String userId) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/reviews/user/$userId'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Review.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load user reviews: ${response.body}');
    }
  }
}